### PR TITLE
MNT, TST: clarify existing behavior of set_montage / __init__ montage

### PR DESCRIPTION
**Pull request by massich**
So far this PR rewrites an existing test to discuss which is the desired behaviour

### CODE REVIEW

**Comment by massich** [comment]
in head~1 works:
```sh
~/code/mne-python eeglab_set_montage
(mne) ❯ 


~/code/mne-python eeglab_set_montage
(mne) ❯ git checkout HEAD~1
Note: checking out 'HEAD~1'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 9537ab550 MNT: Use fixtures to create the file instead of inisde the test

~/code/mne-python 9537ab5...
(mne) ❯ pytest mne/io/eeglab/tests/test_eeglab.py -k position_information
Test session starts (platform: linux, Python 3.6.8, pytest 4.4.0, pytest-sugar 0.9.2)
rootdir: /home/sik/code/mne-python, inifile: setup.cfg
plugins: timeout-1.3.3, sugar-0.9.2, pudb-0.7.0, mock-1.10.3, faulthandler-1.5.0, cov-2.6.1
collecting ... 
 mne/io/eeglab/tests/test_eeglab.py ✓                                                                                                                                               100% ██████████
----------------------------------------------------------------- generated xml file: /home/sik/code/mne-python/junit-results.xml -----------------------------------------------------------------


==================================================================================== slowest 20 test durations ====================================================================================
1.60s setup    mne/io/eeglab/tests/test_eeglab.py::test_position_information
0.01s call     mne/io/eeglab/tests/test_eeglab.py::test_position_information

(0.00 durations hidden.  Use -vv to show these durations.)

Results (11.12s):
       1 passed
      12 deselected
```

current breaks:
```sh
~/code/mne-python 9537ab5... 12s
(mne) ❯ git checkout -                                                                
Previous HEAD position was 9537ab550 MNT: Use fixtures to create the file instead of inisde the test
Switched to branch 'eeglab_set_montage'

~/code/mne-python eeglab_set_montage 12s
(mne) ❯ pytest mne/io/eeglab/tests/test_eeglab.py -k position_information
Test session starts (platform: linux, Python 3.6.8, pytest 4.4.0, pytest-sugar 0.9.2)
rootdir: /home/sik/code/mne-python, inifile: setup.cfg
plugins: timeout-1.3.3, sugar-0.9.2, pudb-0.7.0, mock-1.10.3, faulthandler-1.5.0, cov-2.6.1
collecting ... 

―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_position_information ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

one_chanpos_fname = '/tmp/pytest-of-sik/pytest-326/data0/test_chanpos.set'

    @pytest.mark.filterwarnings('ignore:.*did not have a position.*')
    def test_position_information(one_chanpos_fname):
        """Test reading file with 3 channels - one without position information."""
        from mne.channels import read_montage
        nan = np.nan
        EXPECTED_LOCATIONS_FROM_FILE = np.array([
            [-4.,  1.,  7.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
            [-5.,  2.,  8.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
            [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
        ])
    
        EXPECTED_LOCATIONS_FROM_MONTAGE = np.array([
            [-0.56705965, 0.67706631, 0.46906776, 0, 0, 0, 0, 0, 0, 0, 0, 0],
            [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
            [0, 0.99977915, -0.02101571, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        ])
    
        raw = read_raw_eeglab(input_fname=one_chanpos_fname, preload=True)
        assert_array_equal(np.array([ch['loc'] for ch in raw.info['chs']]),
                           EXPECTED_LOCATIONS_FROM_FILE)
    
        raw = read_raw_eeglab(input_fname=one_chanpos_fname, preload=True,
                              montage=montage)
        _assert_array_equal_nan(np.array([ch['loc'] for ch in raw.info['chs']]),
                                EXPECTED_LOCATIONS_FROM_MONTAGE)
    
        foo = read_raw_eeglab(input_fname=one_chanpos_fname, preload=True)
        foo.set_montage(montage)
    
        # Things that happen:
        montage_info = read_montage(montage)
        for actual, expected in zip([ch['ch_name'] for ch in foo.info['chs']],
                                    ['F3', 'unknown', 'FPz']):
            assert actual == expected
    
        assert_array_equal(foo.info['chs'][0]['loc'][:3],
                           montage_info.pos[montage_info.ch_names.index('F3')])
        assert_array_equal(foo.info['chs'][2]['loc'][:3],
                           montage_info.pos[montage_info.ch_names.index('FPz')])
    
        # There are 3 values but they don't come from montage
        assert_array_equal(foo.info['chs'][1]['loc'][:3], [-5.,  2.,  8.])
        assert np.count_nonzero(
            np.isin(montage_info.pos,
                    np.random.choice(montage_info.pos.flatten(), 3))
        ) == 3
        assert np.count_nonzero(
            np.isin(montage_info.pos, foo.info['chs'][1]['loc'][:3])) == 0
    
        # Things that we would like to happen:
        _assert_array_equal_nan(np.array([ch['loc'] for ch in foo.info['chs']]),
>                               EXPECTED_LOCATIONS_FROM_MONTAGE)

EXPECTED_LOCATIONS_FROM_FILE = array([[-4.,  1.,  7.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [-5.,  2.,  8.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]])
EXPECTED_LOCATIONS_FROM_MONTAGE = array([[-0.56705965,  0.67706631,  0.46906776,  0.        ,  0.        ,
         0.        ,  0.        ,  0.        ....        ,
         0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
         0.        ,  0.        ]])
actual     = 'FPz'
expected   = 'FPz'
foo        = <RawEEGLAB  |  test_chanpos.set, n_channels x n_times : 3 x 3 (0.0 sec), ~12 kB, data loaded>
montage_info = <Montage | test_chans - 32 channels: FPz, EOG1, F3 ...>
nan        = nan
one_chanpos_fname = '/tmp/pytest-of-sik/pytest-326/data0/test_chanpos.set'
raw        = <RawEEGLAB  |  test_chanpos.set, n_channels x n_times : 3 x 3 (0.0 sec), ~12 kB, data loaded>
read_montage = <function read_montage at 0x7fefb2014ae8>

mne/io/eeglab/tests/test_eeglab.py:401: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

left = array([[-0.56705965,  0.67706631,  0.46906776,  0.        ,  0.        ,
         0.        ,  0.        ,  0.        ....        ,
         0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
         0.        ,  0.        ]])
right = array([[-0.56705965,  0.67706631,  0.46906776,  0.        ,  0.        ,
         0.        ,  0.        ,  0.        ....        ,
         0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
         0.        ,  0.        ]])

    def _assert_array_equal_nan(left, right):
        # from mne.utils import _array_equal_nan
        # assert _array_equal_nan(np.array([ch['loc'] for ch in raw.info['chs']]),
        #                         EXPECTED_LOCATIONS_FROM_MONTAGE)
>       assert_array_equal(np.isnan(left), np.isnan(right))
E       AssertionError: 
E       Arrays are not equal
E       
E       (mismatch 33.33333333333333%)
E        x: array([[False, False, False, False, False, False, False, False, False,
E               False, False, False],
E              [False, False, False, False, False, False, False, False, False,...
E        y: array([[False, False, False, False, False, False, False, False, False,
E               False, False, False],
E              [ True,  True,  True,  True,  True,  True,  True,  True,  True,...

left       = array([[-0.56705965,  0.67706631,  0.46906776,  0.        ,  0.        ,
         0.        ,  0.        ,  0.        ....        ,
         0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
         0.        ,  0.        ]])
right      = array([[-0.56705965,  0.67706631,  0.46906776,  0.        ,  0.        ,
         0.        ,  0.        ,  0.        ....        ,
         0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
         0.        ,  0.        ]])

mne/io/eeglab/tests/test_eeglab.py:315: AssertionError

 mne/io/eeglab/tests/test_eeglab.py ⨯                                                                                                                                               100% ██████████
----------------------------------------------------------------- generated xml file: /home/sik/code/mne-python/junit-results.xml -----------------------------------------------------------------


===================================================================================== short test summary info =====================================================================================
FAILED mne/io/eeglab/tests/test_eeglab.py::test_position_information
==================================================================================== slowest 20 test durations ====================================================================================
1.66s setup    mne/io/eeglab/tests/test_eeglab.py::test_position_information
0.01s call     mne/io/eeglab/tests/test_eeglab.py::test_position_information

(0.00 durations hidden.  Use -vv to show these durations.)

Results (11.40s):
       1 failed
         - mne/io/eeglab/tests/test_eeglab.py:350 test_position_information
      12 deselected


```

So does in 0.18:
```sh
~/code/mne-python eeglab_set_montage 13s
(mne) ❯ cp mne/io/eeglab/tests/test_eeglab.py /tmp/test_eeglab.py      

~/code/mne-python eeglab_set_montage
(mne) ❯ git checkout maint/0.18
Switched to branch 'maint/0.18'
Your branch is up to date with 'mne/maint/0.18'.

~/code/mne-python maint/0.18
(mne) ❯ pytest /tmp/test_eeglab.py -k position_information
Test session starts (platform: linux, Python 3.6.8, pytest 4.4.0, pytest-sugar 0.9.2)
rootdir: /tmp
plugins: timeout-1.3.3, sugar-0.9.2, pudb-0.7.0, mock-1.10.3, faulthandler-1.5.0, cov-2.6.1
collecting ... 

―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_position_information ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

one_chanpos_fname = '/tmp/pytest-of-sik/pytest-317/data0/test_chanpos.set'

    @pytest.mark.filterwarnings('ignore:.*did not have a position.*')
    def test_position_information(one_chanpos_fname):
        """Test reading file with 3 channels - one without position information."""
        from mne.channels import read_montage
        nan = np.nan
        EXPECTED_LOCATIONS_FROM_FILE = np.array([
            [-4.,  1.,  7.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
            [-5.,  2.,  8.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
            [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
        ])
    
        EXPECTED_LOCATIONS_FROM_MONTAGE = np.array([
            [-0.56705965, 0.67706631, 0.46906776, 0, 0, 0, 0, 0, 0, 0, 0, 0],
            [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
            [0, 0.99977915, -0.02101571, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        ])
    
        raw = read_raw_eeglab(input_fname=one_chanpos_fname, preload=True)
        assert_array_equal(np.array([ch['loc'] for ch in raw.info['chs']]),
                           EXPECTED_LOCATIONS_FROM_FILE)
    
        raw = read_raw_eeglab(input_fname=one_chanpos_fname, preload=True,
                              montage=montage)
        _assert_array_equal_nan(np.array([ch['loc'] for ch in raw.info['chs']]),
                                EXPECTED_LOCATIONS_FROM_MONTAGE)
    
        foo = read_raw_eeglab(input_fname=one_chanpos_fname, preload=True)
        foo.set_montage(montage)
    
        # Things that happen:
        montage_info = read_montage(montage)
        for actual, expected in zip([ch['ch_name'] for ch in foo.info['chs']],
                                    ['F3', 'unknown', 'FPz']):
            assert actual == expected
    
        assert_array_equal(foo.info['chs'][0]['loc'][:3],
                           montage_info.pos[montage_info.ch_names.index('F3')])
        assert_array_equal(foo.info['chs'][2]['loc'][:3],
                           montage_info.pos[montage_info.ch_names.index('FPz')])
    
        # There are 3 values but they don't come from montage
        assert_array_equal(foo.info['chs'][1]['loc'][:3], [-5.,  2.,  8.])
        assert np.count_nonzero(
            np.isin(montage_info.pos,
                    np.random.choice(montage_info.pos.flatten(), 3))
        ) == 3
        assert np.count_nonzero(
            np.isin(montage_info.pos, foo.info['chs'][1]['loc'][:3])) == 0
    
        # Things that we would like to happen:
        _assert_array_equal_nan(np.array([ch['loc'] for ch in foo.info['chs']]),
>                               EXPECTED_LOCATIONS_FROM_MONTAGE)

/tmp/test_eeglab.py:401: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

left = array([[-0.56705965,  0.67706631,  0.46906776,  0.        ,  0.        ,
         0.        ,  0.        ,  0.        ....        ,
         0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
         0.        ,  0.        ]])
right = array([[-0.56705965,  0.67706631,  0.46906776,  0.        ,  0.        ,
         0.        ,  0.        ,  0.        ....        ,
         0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
         0.        ,  0.        ]])

    def _assert_array_equal_nan(left, right):
        # from mne.utils import _array_equal_nan
        # assert _array_equal_nan(np.array([ch['loc'] for ch in raw.info['chs']]),
        #                         EXPECTED_LOCATIONS_FROM_MONTAGE)
>       assert_array_equal(np.isnan(left), np.isnan(right))
E       AssertionError: 
E       Arrays are not equal
E       
E       (mismatch 33.33333333333333%)
E        x: array([[False, False, False, False, False, False, False, False, False,
E               False, False, False],
E              [False, False, False, False, False, False, False, False, False,...
E        y: array([[False, False, False, False, False, False, False, False, False,
E               False, False, False],
E              [ True,  True,  True,  True,  True,  True,  True,  True,  True,...

/tmp/test_eeglab.py:315: AssertionError

 test_eeglab.py ⨯                                                                                                                                                                   100% ██████████

Results (0.46s):
       1 failed
         - test_eeglab.py:350 test_position_information
      12 deselected


~/code/mne-python eeglab_set_montage 12s
(mne) ❯ 
```

File: mne/io/eeglab/tests/test_eeglab.py [hunk 1]
```diff
@@ -308,4 +308,92 @@ def test_eeglab_event_from_annot():
...
+    # a mix of what is in montage and in the file
+
+    # flushing
+    foo = read_raw_eeglab(input_fname=one_chanpos_fname, preload=True)
+    foo.set_montage(None)
+    foo.set_montage(montage)
+    _assert_array_equal_nan(np.array([ch['loc'] for ch in foo.info['chs']]),
+                            EXPECTED_LOCATIONS_FROM_MONTAGE)
+
+    # flushing
```
Comment (agramfort) [inline mne/io/eeglab/tests/test_eeglab.py]: why do you say flushing here?

File: mne/io/eeglab/tests/test_eeglab.py [hunk 2]
```diff
@@ -308,4 +308,92 @@ def test_eeglab_event_from_annot():
     assert len(events_b) == 154
 
 
+def _assert_array_equal_nan(left, right):
+    # from mne.utils import _array_equal_nan
+    # assert _array_equal_nan(np.array([ch['loc'] for ch in raw.info['chs']]),
+    #                         EXPECTED_LOCATIONS_FROM_MONTAGE)
```
Comment (agramfort) [inline mne/io/eeglab/tests/test_eeglab.py]: cleanup?

**Comment by massich** [comment]
I guess it can go in. But my main goal with this PR was to unfold the behavior of the set_montage.

**Pull request merged**